static/js/lib/redux_api.js

```diff
@@ -0,0 +1,166 @@
+
+  const requestType = requestActionType(prefix, endpoint.name);
+  const successType = successActionType(prefix, endpoint.name);
+  const failureType = failureActionType(prefix, endpoint.name);
+
+  const requestAction = createAction(requestType);
+  const successAction = createAction(successType);
+  const failureAction = createAction(failureType);
+
+  const fetchFunc = R.propOr(
+    makeFetchFunc(endpoint),
+    `${R.toLower(verb)}Func`,
+    endpoint
+  );
+
+  return {
+    action: (...params) => {
+      return (dispatch: Dispatch): Promise<*> => {
+        dispatch(requestAction());
+        return fetchFunc(...params).then(data => {
+          dispatch(successAction(data));
+          return Promise.resolve(data);
+        }).catch(error => {
+          dispatch(failureAction(error));
+          return Promise.reject(error);
+        });
+      };
+    },
+    requestType, successType, failureType
+  };
+};
+
+// here we loop through the HTTP verbs that an Endpoint implements, and
+// for each one we derive an action. The derived action types are stuck on
+// the function object, as a convenience.
+export const deriveActions = (endpoint: Endpoint) => {
+  let actions = {};
+  endpoint.verbs.forEach(verb => {
+    const {
+      action,
+      requestType,
+      successType,
+      failureType,
+    } = deriveAction(endpoint, verb);
+
+    let lverb = R.toLower(verb);
+
+    actions[lverb] = action;
+    actions[lverb].requestType = requestType;
+    actions[lverb].successType = successType;
+    actions[lverb].failureType = failureType;
+  });
+  let clearType = clearActionType(endpoint.name);
+  actions.clearType = clearType;
+  actions.clear = createAction(clearType);
+  return actions;
+};
+
+// a reducer may be derived on the basis of an Endpoint object, an already
+// derived REST action, and an HTTP verb
+// A reducer, in this case, is an Object<type, Function>, where type is an
+// action type defined on the previously derived action, and the Function has
+// the type State -> Action -> State
+// 
+// so the type of the function overall could be rendered
+// deriveReducer :: Endpoint -> Action -> String -> Object String (State -> Action -> State)
+export const deriveReducer = (endpoint: Endpoint, action: Function, verb: string) => {
+  let fetchStatus = `${R.toLower(verb)}Status`;
+
+  return { 
+    [action.requestType]: (state: Object, action: Action<any, any>) => ({
+      ...state,
+      [fetchStatus]: FETCH_PROCESSING,
+      loaded: false,
+      processing: true,
+    }),
+    [action.successType]: (state: Object, action: Action<any, any>) => ({
+      ...state,
+      [fetchStatus]: FETCH_SUCCESS,
+      data: action.payload,
+      loaded: true,
+      processing: false,
+    }),
+    [action.failureType]: (state: Object, action: Action<any, any>) => ({
+      ...state,
+      [fetchStatus]: FETCH_FAILURE,
+      error: action.payload,
+      loaded: true,
+      processing: false,
+    }),
+  };
```

Comment:
Yes, before I was defining two redundant handlers (one for each HTTP verb) which didn't make much sense to me. I refactored it a little bit to define one clear action for the whole reducer. So now the action type and action creator are defined in `deriveActions` instead of `deriveAction` (the latter being the per-verb derivation function) and the function to handle the clear action in the reducer is now defined in `deriveReducers` instead of `deriveReducer` (which, again, is the per-verb function).

The relevant code is on line 147, adding the handler once for the whole reducer. It also allows the creator of the `Endpoint` object to override the default `INITIAL_STATE` if that is necessary (by putting an `initialState` object on the `Endpoint`).